File: 30-Day LeetCoding Challenge/Python/stringShift.py
class Solution(object):
    def stringShift(self, s, shift):
        """
        :type s: str
        :type shift: List[List[int]]
        :rtype: str
        """
        step = 0
        length = len(s)
        is_left = False
        for el in shift:
            if el[0] == 0:
                step -= el[1]
            else:
                step += el[1]
        if 0 == step:
            return s
        elif step > 0:
            is_left = False
        else:
            is_left = True
        step = abs(step)%length
        s_after_shift = list(s)
        for i in range(length):
            if is_left:
                if i >= step:
                    s_after_shift[i-step] = s[i]
                else:
                    s_after_shift[length-step+i] = s[i]
            else:
                if i + step > length - 1:
                    s_after_shift[i + step - length] = s[i]
                else:
                    s_after_shift[i+ step] = s[i]
        return "".join(s_after_shift)
                

input = [[1,1],[1,1],[0,2],[1,3]]
str = "abcdefg"
s = Solution()
print(s.stringShift(str,input))

File: 30-Day LeetCoding Challenge/C++/stringShift.cpp
#include <iostream>
#include <algorithm>
#include <string>
#include <vector>

using namespace std;

class Solution {
public:
    string stringShift(string s, vector<vector<int>>& shift) {
        int length = s.length();
        int x = 0;
        bool is_left;
        for(int i = 0; i < shift.size(); i++) {
            if(shift[i][0] == 0) x -= shift[i][1];
            else x += shift[i][1];
        }
        if(x == 0) return s;
        else if(x > 0) is_left = false;
        else is_left = true;
        x = abs(x)%length;
        string s_after_shift = s;
        for(int i = 0; i < length; i++) {
            if(is_left) {
                if(i >= x) s_after_shift[i-x] = s[i];
                else s_after_shift[length-x+i] = s[i];
            }
            else {
                if(i + x > length - 1) s_after_shift[i + x - length] = s[i];
                else s_after_shift[i + x] = s[i];
            }
        }
        return s_after_shift;
    }
};

int main() {
    Solution s;
    vector<vector<int>> input = {{1,1},{1,1},{0,2},{1,3}};
    string str = "abcdefg";
    cout << s.stringShift(str,input);
    return 0;

}